Add notation sign CLI spec (#341)
notaryproject/notation: specs/commandline/sign.md

# notation sign
## Description
Use `notation sign` to sign artifacts.

> If the container image being signed is stored in the registry, upon successful signing, the generated signature will be pushed to the registry and the digest of the container image will be returned.

Author: Agree and thanks for the tip.

## Outline
```console
Sign artifacts

Usage:
  notation sign <reference> [flags]

> Flags:

Author: Agree.

      --cert-file string        Location of file containing signing(leaf) certificate and certificate chain

>   -e, --expiry duration         Expire duration in seconds, minutes or hours

Author: Agree

  -h, --help                    Help for sign

>   -k, --key string              Signing key name

Author: Agree.

      --key-file string         Signing key file

>   -p, --password string         Password for registry operations (default from $NOTATION_PASSWORD)

Author: Agree

  -c, --pluginConfig string     List of comma-separated {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values
  -u, --username string         Username for registry operations (default from $NOTATION_USERNAME)

Global Flags:

>       --plain-http   Registry access via plain HTTP

Author: Since it is a global flag, I'd suggest tracking this in another issue or PR.

```
## Usage
### Sign a container image
```console
# Add a key and make it a default signing key

> notation key add -n <key name> <key path> <cert path> --default

Author: Yes, key name, I will use a full name to replace the short one '-n'

# [Optional] Change a default signing key
notation key update <key name> --default

# Sign a container image using the default signing key

> notation sign <image>

Author: Agree.

```
### Sign a container image using a remote key
```console
# Prerequisites: 

> # - A Key Vault plugin is installed in notation

Author: Agree. Thanks.

> # - User creates keys and certificates in a Key Vault

Author: Agree.

> # Add a default signing key referencing the key stored in the Key Vault

Author: Agree

notation key add -n <key name> --plugin <plugin name> --id <remote key id> --default

# sign a container image using a remote key
notation sign <image>
```
### Sign a container image and specify the signature expiry duration, for example 24 hours
```console

> notation sign <image> --expiry 24h

Author: User cannot specify duration in day. I will change `for example 24 hours` to `for example 1 day`, but keep the `--expiry 24h`.

# notation sign
## Description
Use `notation sign` to sign artifacts.

Signs a container artifact that is stored in a registry. Upon successful signing, the generated signature is pushed to the registry and the digest of the container image is returned.
## Outline
```
Sign artifacts

Usage:
  notation sign <reference> [flags]

Flags:

>       --cert-file string        Location of file containing signing(leaf) certificate and certificate chain. Use this flag with flag '--key-file' together.

Reviewer: ```suggestion
      --cert-file string        Location of file containing signing(leaf) certificate and certificate chain. Use this flag with '--key-file' .
```

>   -e, --expiry duration         Optional expiry that provides a “best by use” time for the artifact. The duration is specified in seconds, minutes or hours.

Reviewer: IMO there is no real world usecase where customer would want an artifact to expire in second.
```suggestion
  -e, --expiry duration         Optional expiry that provides a “best by use” time for the artifact. The duration is specified in minutes, hours or days.
```

>   -e, --expiry duration         Optional expiry that provides a “best by use” time for the artifact. The duration is specified in seconds, minutes or hours.

Reviewer 2: Let's keep ASCII.
```suggestion
  -e, --expiry duration         Optional expiry that provides a "best by use" time for the artifact. The duration is specified in seconds, minutes or hours.
```

  -h, --help                    Help for sign
  -k, --key string              Signing key name, for a key previously added to notation's key list.

>       --key-file string         Signing key file. Use this flag with flag '--cert-file' together.

Reviewer: ```suggestion
      --key-file string         Location of file containing signing key file. Use this flag with  '--cert-file'.
```

  -p, --password string         Password for registry operations (default to $NOTATION_PASSWORD if not specified)

>   -c, --pluginConfig string     Optional list of comma-separated {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values

Reviewer: For consistency we should use snake case everywhere.
```suggestion
  -c, --plugin-config string     Optional, list of comma-separated {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values
```

  -u, --username string         Username for registry operations (default to $NOTATION_USERNAME if not specified)

Global Flags:
      --plain-http   Registry access via plain HTTP
```
## Usage
### Sign a container image
```
# Add a key which uses a local private key and certificate, and make it a default signing key
notation key add --name <key name> <key path> <cert path> --default

# [Optional] Change a default signing key
notation key update <key name> --default

# Sign a container image using the default signing key
notation sign https://<registry>/<repository>:<tag>
```
### Sign a container image using a remote key
```
# Prerequisites: 

> # - A compliant signing plugin is installed in notation. See notation plugin documentation for more details (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md).

Reviewer: ```suggestion
# - A compliant signing plugin is installed in notation. See notation [plugin documentation](https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.
```

Author: This link is under a block, so it won't work using format `[](https://)`

---
title: "notation sign"
description: "The notation sign command description and usage"

> keywords: "notation, sign"

Reviewer 3: I think this header is not necessary, we can remove this for now.

Author: Agree. We can remove it untill there is a valid usage.

---
# notation sign
## Description
Use `notation sign` to sign artifacts.

If the signing artifact is a container image stored in a registry, the signature is pushed to the registry by default after signing successfully, and the digest of the container image is returned.
## Outline
```console
$ notation sign --help
Sign artifacts

Usage:
  notation sign [reference] [flags]

Flags:
      --cert-file string        signing certificate file

>   -e, --expiry duration         expire duration

Author: No, it can be ns, us, ms, s, m, or h. I can add this info in the description.

  -h, --help                    help for sign
  -k, --key string              signing key name
      --key-file string         signing key file

>   -l, --local                   reference is a local file

Reviewer 3: Can we explicitly claim this flag needs to be used with `-o` together?

Author: Second thought: We can remove both "-l" and "-o" flag for rc.1, because we don't have a complete use story implemented for signing local artifacts. It just works partially. We can add it back once we have a complete user story implemented.

>       --media-type string       specify the media type of the manifest read from file or stdin (default "application/vnd.docker.distribution.manifest.v2+json")

Author: We don't support signing local images, but we can sign a local file, for example a SBOM file, and the signature is stored locally.

  -o, --output string           write signature to a specific path
  -p, --password string         Password for registry operations (default from $NOTATION_PASSWORD)
  -c, --pluginConfig string     list of comma-separated {key}={value} pairs that are passed as is to the plugin, refer plugin documentation to set appropriate values

>       --push                    push after successful signing (default true)

Author: I will vote for this suggestion. let's wait for more comments.

>       --push                    push after successful signing (default true)

Reviewer 3: +1. Push signature is a default behavior after successful signing, which means `--no-push`  is much more straightforward than `--push boolean`.

>       --push                    push after successful signing (default true)

Author: Second thought: if there is no such use case "not pushing signature to remote" for rc.1, I perfer to remove this flag. We can add it back when there is a valid use case. Once we release this flag, it is not easy to remove it.

>   -r, --reference string        original reference

Author: This flag is useless, and I will remove it.

  -u, --username string         Username for registry operations (default from $NOTATION_USERNAME)

Global Flags:
      --plain-http   Registry access via plain HTTP
```

> ## Usage

Reviewer 3: I strongly recommend we add an _Example_ section or paragraph for each usage. The _Example_ for each command is currently missing in the Notation help doc, maybe we can start to prepare it from here.

Author: As discussed, we will not add `Example` in the spec, it will be added in the command help doc.  I will remove the line of `$notation sign --help`, which will cause confusion that it is an output of `notation sign` command.

### sign a container image with a local key and certificate
```console
notation sign <image> --key-file <key path> --cert-file <cert path> 
```
### sign a container image using a key name
```console
# Add a key with a key name referencing signing key file
notation key add -n <key name> <key path> <cert path> 

# sign a container image using a key name
notation sign <image> --key <key name>
```
### sign a container image with key and certificate stored in a Key Vault
```console

> # Pre-condition: 

Reviewer 3: How about updating `Pre-condition` to `Prerequisites`?

Author: Agree.

# - A Key Vault plugin is installed in notation
# - User creates keys and certificates in a Key vault
# Add the key with a key name referencing the key stored in Key Vault
notation key add -n <key name> --plugin <plugin name> --id <key id>

> # sign a container image using a key name

Reviewer 3: The first letter needs to be capitalized.

notation sign <image> --key <key name>
```

> ### store signature in a local file

Reviewer 3: Same as above

# notation sign

## Description

Use `notation sign` to sign artifacts.

Signs a container artifact that is stored in a registry. Upon successful signing, the generated signature is pushed to the registry and the digest of the container image is returned.

## Outline

```text
Sign artifacts

Usage:
  notation sign <reference> [flags]

Flags:
      --cert-file string        Location of file containing signing(leaf) certificate and certificate chain. Use this flag with '--key-file'.

>   -e, --expiry duration         Optional expiry that provides a "best by use" time for the artifact. The duration is specified in minutes, hours or days.

Reviewer 4: The expiry will be in terms of days/months or years. We should probably take a expiry date and time instead of duration here.

>   -e, --expiry duration         Optional expiry that provides a "best by use" time for the artifact. The duration is specified in minutes, hours or days.

Author: @gokarnm the usability is not good if user is asked to input date and time like `2023/01/02T05:30:12Z`, so I would like to keep the expiry in a duration format.

  -h, --help                    Help for sign
  -k, --key string              Signing key name, for a key previously added to notation's key list.
      --key-file string         Location of file containing signing key file. Use this flag with '--cert-file'.
  -p, --password string         Password for registry operations (default to $NOTATION_PASSWORD if not specified)

>   -c, --plugin-config string     Optional, list of comma-separated {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values

Reviewer: ```suggestion
  -c, --plugin-config string    Optional, list of comma-separated {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values
```

  -u, --username string         Username for registry operations (default to $NOTATION_USERNAME if not specified)

Global Flags:
      --plain-http   Registry access via plain HTTP
```

## Usage

### Sign a container image

```text
# Add a key which uses a local private key and certificate, and make it a default signing key
notation key add --name <key name> <key path> <cert path> --default

# [Optional] Change a default signing key
notation key update <key name> --default

# Sign a container image using the default signing key

> notation sign https://<registry>/<repository>:<tag>

Reviewer 2: Reference does not have a URI scheme
```suggestion
notation sign <registry>/<repository>:<tag>
```

Additionally, it is more secure to do
```shell
notation sign <registry>/<repository>@<digest>
```

# notation sign

## Description

Use `notation sign` to sign artifacts.

Signs a container artifact that is stored in a registry. Upon successful signing, the generated signature is pushed to the registry and the digest of the container image is returned.

## Outline

```text
Sign artifacts

Usage:
  notation sign <reference> [flags]

Flags:
      --cert-file string        Location of file containing a complete certificate chain for the signing key. Use this flag with '--key-file'.
  -e, --expiry duration         Optional expiry that provides a "best by use" time for the artifact. The duration is specified in minutes(m), hours(h) or days(d). For example: 30d, 12h, 30m, 1d3h20m
  -h, --help                    Help for sign
  -k, --key string              Signing key name, for a key previously added to notation's key list.
      --key-file string         Location of file containing signing key file. Use this flag with '--cert-file'.

>   -p, --password string         Password or identity token for registry operations (default to $NOTATION_PASSWORD if not specified)
>   --plugin-config strings       List of {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values
>   -u, --username string         Username for registry operations (default to $NOTATION_USERNAME if not specified)

Reviewer 2: nit: Indentation.
```suggestion
  -p, --password string         Password or identity token for registry operations (default to $NOTATION_PASSWORD if not specified)
      --plugin-config strings   List of {key}={value} pairs that are passed as is to a plugin, if the key (--key) is associated with a signing plugin, refer plugin documentation to set appropriate values
  -u, --username string         Username for registry operations (default to $NOTATION_USERNAME if not specified)
```

Global Flags:
      --plain-http   Registry access via plain HTTP
```

## Usage

> ### Sign a container image

Reviewer 2: Should we mention `artifact` instead of container image in all the places?

```shell
# Add a key which uses a local private key and certificate, and make it a default signing key

> notation key add --name <key_name> <key_path> <cert_path> --default

Reviewer 2: It is better to put all flags before arguments although they can be put after the arguments.

# Or change the default signing key to an existing signing key
notation key update <key_name> --default

# Sign a container image using the default signing key
notation sign <registry>/<repository>:<tag>

# Or using container image digests instead of tags. A container image digest uniquely and immutably identifies a container image.
notation sign <registry>/<repository>@<digest>
```

### Sign a container image using a remote key

```shell
# Prerequisites: 
# - A compliant signing plugin is installed in notation. See notation plugin documentation (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.
# - User creates keys and certificates in a 3rd party key provider (e.g. key vault, key management service). The signing plugin installed in previous step must support generating signatures using this key provider.

# Add a default signing key referencing the key identifier for the remote key, and the plugin associated with it.

> notation key add --name <key_name> --plugin <plugin name> --id <remote key id> --default

Reviewer 2: ```suggestion
notation key add --name <key_name> --plugin <plugin_name> --id <remote_key_id> --default
```

# - User creates keys and certificates in a 3rd party key provider (e.g. key vault, key management service). The signing plugin installed in previous step must support generating signatures using this key provider.

# Add a default signing key referencing the key identifier for the remote key, and the plugin associated with it.
notation key add --default --name <key_name> --plugin <plugin_name> --id <remote_key_id>

# sign a container image using a remote key
notation sign <registry>/<repository>:<tag>
```

### Sign an OCI artifact using the default signing key

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use a digest that uniquely and immutably identifies an OCI artifact.
notation sign <registry>/<repository>@<digest>
```

### Sign a container image and specify the signature expiry duration, for example 1 day

```shell
notation sign --expiry 1d <registry>/<repository>:<tag>
```

### Sign a container image using a specified signing key

```shell
# List signing keys to get the key name
notation key list

# Sign a container image using the specified key name

> notation sign <registry>/<repository>:<tag> --key <key_name>

Reviewer 2: ```suggestion
notation sign --key <key_name> <registry>/<repository>:<tag>
```

```

### Sign a container image using a local key and certificate which are not added in the signing key list

```shell

> notation sign <registry>/<repository>:<tag> --key-file <key_path> --cert-file <cert_path>

Reviewer 2: ```suggestion
notation sign --key-file <key_path> --cert-file <cert_path> <registry>/<repository>:<tag>
```
